tests/YahooCurrencyRateProviderTest.php

```diff
@@ -0,0 +1,210 @@
+
+        $this->assertEquals(2, count($rates));
+        foreach($rates as $rate) {
+            $this->assertInstanceOf('\\RedCode\\Currency\\Rate\\ICurrencyRate', $rate);
+        }
+    }
+
+    public function testYahooCurrencyRateProviderGetRatesForIncorrectDate()
+    {
+        $currencies = [];
+        $currencies['EUR'] = $this->getMock('\\RedCode\\Currency\\ICurrency');
+        $currencies['EUR']
+            ->method('getCode')
+            ->willReturn('EUR');
+
+        $currencies['RUB'] = $this->getMock('\\RedCode\\Currency\\ICurrency');
+        $currencies['RUB']
+            ->method('getCode')
+            ->willReturn('RUB')
+        ;
+
+        $currentDate = new \DateTime('now');
+        $estDate = new \DateTime('now', new \DateTimeZone('EST'));
+
+        try {
+            $this->currencyRateProvider->getRates(array_values($currencies));
+        } catch (\Exception $e) {
+            if ((in_array($currentDate->format('w'), ['6', '7'], true)) ||
+                ($currentDate->format('Y-m-d') !== $estDate->format('Y-m-d'))) {
+                $this->assertInstanceOf('\\RedCode\\Currency\\Rate\\Exception\\NoRatesAvailableForDateException', $e);
+            }
+        }
+    }
+
+    /**
+     * @expectedException \RedCode\Currency\Rate\Exception\BadXMLQueryException
+     * @expectedExceptionMessageRegExp #Could not create XML from query ".*" for provider yahoo#
+     */
+    public function testYahooCurrencyRateProviderGetRatesWithBadXML()
+    {
+        $currencyManager = $this->getMock('\\RedCode\\Currency\\ICurrencyManager');
+        $currencyRateManager = $this->getMock('\\RedCode\\Currency\\Rate\\ICurrencyRateManager');
+
+        $xmlLoader = $this->getMock('\\RedCode\\Currency\\Rate\\XML\\XMLLoader');
+        $xmlLoader
+            ->method('load')
+            ->willReturn(false);
+
+        $this->currencyRateProvider = new YahooCurrencyRateProvider(
+            $currencyRateManager,
+            $currencyManager,
+            $xmlLoader
+        );
+
+        $currencies = [];
+        $currencies['EUR'] = $this->getMock('\\RedCode\\Currency\\ICurrency');
+        $currencies['EUR']
+            ->method('getCode')
+            ->willReturn('EUR');
+
+        $this->currencyRateProvider->getRates(array_values($currencies));
+    }
+
+    /**
+     * @expectedException \RedCode\Currency\Rate\Exception\NoRatesAvailableForDateException
+     * @expectedExceptionMessageRegExp #No rates available for ....-..-.. date with provider yahoo#
+     */
+    public function testYahooCurrencyRateProviderGetRatesWithNullObject()
+    {
+        $currencyManager = $this->getMock('\\RedCode\\Currency\\ICurrencyManager');
+        $currencyRateManager = $this->getMock('\\RedCode\\Currency\\Rate\\ICurrencyRateManager');
+
+        $quote = (object)['quote' => []];
+        $xmlResponse = (object)['results' => $quote];
```

Comment:
You can add methods in the tests, which return mocked XMLLoader with 
- correct response and 
- incorrect response 

And use them in places where that is needed.
Do the same for Ecb porvider